### DISCUSSION TITLE
Disable _deps folder for cppcheck

### DISCUSSION THREAD

**Discussion by huand** [TheLartians/Greeter#76] [Q&A]
Hello!
I guess a simple question but I can't seem to make it work.
How to exclude the _deps folder from cpm.cmake from being checked with cppcheck?

my configure command:

`/usr/bin/cmake --no-warn-unused-cli -DUSE_CCACHE=ON -DUSE_STATIC_ANALYZER=cppcheck -DCPPCHECK_ARGS=--enable=all -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug -DCMAKE_C_COMPILER:FILEPATH=/bin/gcc-9 -DCMAKE_CXX_COMPILER:FILEPATH=/bin/g++-9 -H/home/hub/Downloads/ModernCppStarter-master/test -B/home/hub/Downloads/ModernCppStarter-master/build -G Ninja`

example of output (check doctest in _deps):

```
[build] _deps/doctest-src/doctest/doctest.h:4894:9: warning: Member variable 'ConsoleReporter::hasLoggedCurrentTestStart' is not initialized in the constructor. [uninitMemberVar]
[build]         ConsoleReporter(const ContextOptions& co, std::ostream& ostr)
[build]         ^
[build] _deps/doctest-src/doctest/doctest.h:4894:9: warning: Member variable 'ConsoleReporter::tc' is not initialized in the constructor. [uninitMemberVar]
[build]         ConsoleReporter(const ContextOptions& co, std::ostream& ostr)
[build]         ^

...
```

i tried the -i _deps  option without success. any idea what i am missing?
thx for all the tools!

**Comment by TheLartians** [comment]
Not a user of cppcheck myself, so these are only guesses, but you could try:

- Defining the [`CPM_SOURCE_CACHE`](https://github.com/TheLartians/CPM.cmake#cpm_source_cache) environmental variable to an external directory (e.g. to `$HOME/.cache/cpm`) so that the deps will be cached and stored outside the build directory
- Did you pass the `-i` argument inside the `CPPCHECK_ARGS` variable (something like `-DCPPCHECK_ARGS="--enable=all;-ipath/to/_deps"`)? Only then they will be forwarded to cppcheck.

**Reply by huand** [reply]
couldnt make it work i guess ill stick to use cppcheck outside of cmake it's not really  an issue

**Comment by ClausKlein** [comment]
I have a related issue with `clang-tidy` https://github.com/TheLartians/ModernCppStarter/issues/89

But I use `CPM_SOURCE_CACHE`

**Reply by ClausKlein** [reply]
@TheLartians would this be a solution/workaround?
https://discourse.cmake.org/t/how-to-prevent-clang-tidy-to-check-sources-added-with-fetchcontents/2871

**Reply by TheLartians** [reply]
I guess if you add the code around any `CPMAddPackage` calls it should work.

**Reply by ClausKlein** [reply]
unfortunately it doesn't help
```diff
Claus-iMac:cpm clausklein$ diff CPM_0.31.1.cmake CPM_0.31.1.cmake.ck -u 
--- CPM_0.31.1.cmake	2021-03-03 10:11:44.586209207 +0100
+++ CPM_0.31.1.cmake.ck	2021-03-03 10:14:19.849974978 +0100
@@ -703,6 +703,7 @@
   if(${CPM_DRY_RUN})
     return()
   endif()
+
   FetchContent_GetProperties(${PACKAGE})
   string(TOLOWER ${PACKAGE} lpackage)
   set(${PACKAGE}_SOURCE_DIR
@@ -721,7 +722,25 @@
   else()
     set(addSubdirectoryExtraArgs "")
   endif()
+
+  foreach (lang IN ITEMS C CXX)
+    if(DEFINED "CMAKE_${lang}_CLANG_TIDY")
+      message(WARNING "unset CMAKE_${lang}_CLANG_TIDY before add_subdirectory(${SOURCE_DIR})")
+      set("CMAKE_${lang}_CLANG_TIDY_save" "${CMAKE_${lang}_CLANG_TIDY}")
+      unset("CMAKE_${lang}_CLANG_TIDY")
+    endif()
+  endforeach ()
+
   add_subdirectory(${SOURCE_DIR} ${BINARY_DIR} ${addSubdirectoryExtraArgs})
+
+  foreach (lang IN ITEMS C CXX)
+    if(DEFINED "CMAKE_${lang}_CLANG_TIDY_save")
+      set("CMAKE_${lang}_CLANG_TIDY" "${CMAKE_${lang}_CLANG_TIDY_save}")
+      unset("CMAKE_${lang}_CLANG_TIDY_save")
+      message(WARNING "CMAKE_${lang}_CLANG_TIDY=${CMAKE_${lang}_CLANG_TIDY} restored")
+    endif()
+  endforeach ()
+
 endfunction()
 
 # downloads a previously declared package via FetchContent
Claus-iMac:cpm clausklein$ 
```

**Comment by ClausKlein** [comment]
@TheLartians the only way I know so far is to patch the cmake generated file `compile_commands.json`: 
https://github.com/TheLartians/ModernCppStarter/blob/3d006248f5e41729198d7657e791edf787a436b7/GNUmakefile#L38
and than i.e.
https://github.com/TheLartians/ModernCppStarter/blob/3d006248f5e41729198d7657e791edf787a436b7/GNUmakefile#L39
**cppcheck** can work with this file too!

I prefer to use `clang-tidy` in this way.

**Comment by ClausKlein** [comment]
With `CMake v3.25` and `CPM.cmake v0.38` the SYSTEM option may be used.

How this helps shows may project: [ModernCmakeStarter](https://github.com/ClausKlein/ModernCmakeStarter/pull/7)
